Assets/MixedRealityToolkit/Utilities/Rendering/MaterialInstance.cs

```diff
@@ -0,0 +1,305 @@
+
+namespace Microsoft.MixedReality.Toolkit.Rendering
+{
+    /// <summary>
+    /// The MaterialInstance behavior aides in tracking instance material lifetime and automatically destroys instanced materials for the user. 
+    /// This utility component can be used as a replacement to <see href="https://docs.unity3d.com/ScriptReference/Renderer-material.html">Renderer.material</see> or 
+    /// <see href="https://docs.unity3d.com/ScriptReference/Renderer-materials.html">Renderer.materials</see>. When invoking Unity's Renderer.material(s), Unity 
+    /// automatically instantiates new materials. It is the caller's responsibility to destroy the materials when a material is no longer needed or the game object is 
+    /// destroyed. The MaterialInstance behavior helps avoid material leaks and keeps material allocation paths consistent during edit and run time.
+    /// </summary>
+    [HelpURL("https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Rendering/MaterialInstance.html")]
+    [ExecuteAlways, RequireComponent(typeof(Renderer))]
+    public class MaterialInstance : MonoBehaviour
+    {
+        /// <summary>
+        /// Returns the first instantiated Material assigned to the renderer, similar to <see href="https://docs.unity3d.com/ScriptReference/Renderer-material.html">Renderer.material</see>. 
+        /// If any owner is specified the instanced material(s) will not be released until all owners are released. When a material
+        /// is no longer needed ReleaseMaterial should be called with the matching owner.
+        /// </summary>
+        /// <param name="owner">An optional owner to track instance ownership.</param>
+        /// <returns>The first instantiated Material.</returns>
+        public Material AcquireMaterial(object owner = null, bool instance = true)
+        {
+            if (owner != null)
+            {
+                materialOwners.Add(owner);
+            }
+
+            if (instance)
+            {
+                AcquireInstances();
+            }
+
+            if (instanceMaterials?.Length > 0)
+            {
+                return instanceMaterials[0];
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns all the instantiated materials of this object, similar to <see href="https://docs.unity3d.com/ScriptReference/Renderer-materials.html">Renderer.materials</see>. 
+        /// If any owner is specified the instanced material(s) will not be released until all owners are released. When a material
+        /// is no longer needed ReleaseMaterial should be called with the matching owner.
+        /// </summary>
+        /// <param name="owner">An optional owner to track instance ownership.</param>
+        /// <param name="instance">Should this acquisition attempt to instance materials?</param>
+        /// <returns>All the instantiated materials.</returns>
+        public Material[] AcquireMaterials(object owner = null, bool instance = true)
```

Comment:
It might make sense to limit owners to UnityEngine.Objects instead of a generic objects. It would leave the door open for auto-detection of dead owners / auto-release on destroy / stuff like that.

Reply:
I like that idea! I'll change that.